src/hardware/mixer.cpp

```diff
@@ -874,19 +902,35 @@ void MIXER_Init(Section* sec) {
 		mixer.tick_add=calc_tickadd(mixer.freq);
 		TIMER_AddTickHandler(MIXER_Mix_NoSound);
 	} else {
-		if((static_cast<Bit16s>(mixer.freq) != obtained.freq) || (mixer.blocksize != obtained.samples))
-			LOG_MSG("MIXER: Got different values from SDL: freq %d, blocksize %d",obtained.freq,obtained.samples);
-		mixer.freq=obtained.freq;
-		mixer.blocksize=obtained.samples;
-		mixer.tick_add=calc_tickadd(mixer.freq);
+		// Does SDL want a different playback rate?
+		assert(obtained.freq > 0 &&
+		       static_cast<unsigned>(obtained.freq) < UINT32_MAX);
+		const auto obtained_freq = static_cast<uint32_t>(obtained.freq);
+		if (obtained_freq != mixer.freq) {
+			LOG_MSG("MIXER: SDL changed the playback rate from %d to %d Hz",
+			        mixer.freq, obtained_freq);
+			mixer.freq = obtained_freq;
+		}
+
+		// Does SDL want a different blocksize?
+		const auto obtained_blocksize = obtained.samples;
+		if (obtained_blocksize != mixer.blocksize) {
+			LOG_MSG("MIXER: SDL changed the blocksize from %u to %u bytes",
+			        mixer.blocksize, obtained_blocksize);
+			mixer.blocksize = obtained_blocksize;
+		}
+		mixer.tick_add = calc_tickadd(mixer.freq);
 		TIMER_AddTickHandler(MIXER_Mix);
 		SDL_PauseAudioDevice(mixer.sdldevice, 0);
+
+		LOG_MSG("MIXER: Negotiated a %d-Hz output rate and %u-byte blocksize",
+		        mixer.freq, mixer.blocksize);
 	}
-	mixer.min_needed=section->Get_int("prebuffer");
-	if (mixer.min_needed>100) mixer.min_needed=100;
-	mixer.min_needed=(mixer.freq*mixer.min_needed)/1000;
-	mixer.max_needed=mixer.blocksize * 2 + 2*mixer.min_needed;
-	mixer.needed=mixer.min_needed+1;
+	const auto requested_prebuffer = section->Get_int("prebuffer");
+	mixer.min_needed = static_cast<uint16_t>(clamp(requested_prebuffer, 0, 100));
```

Comment:
`clamp` arguments in this case are all signed, so it's output is also signed which generates a sign-conversion warning on assignment.

I'm using a `uint16_t` because that's really what the value should be stored as, because it's value makes up a portion of the fixed 16 KB audio buffer (so capping it at ~64 KiB is pretty natural).